### DISCUSSION TITLE
vcpkg/cmake can't seem to find the Windows 10 SDK

### DISCUSSION THREAD

**Discussion by michaeladamkatz** [microsoft/vcpkg#31934] [Q&A]
I have been stuck on this for several days.

I have a home computer and a work computer. I can install packages on my home computer. I can't install packages on my work computer.

By looking at the logs, I can see that the build on my home computer uses the Windows SDK correctly, at `C:\Program Files (x86)\Windows Kits\10`. On my work computer, it uses the binaries in `C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.36.32532\bin`. (This causes various problems. The first problem was that `mt.exe` and `rc.exe` were not in that `bi`n folder, in either x86 or x64. This led to various experiments of me copying such binaries from various SDK folders into that `bin` folder. That led to other problems, which I can detail, but at the moment I'm focused on just trying to get vcpkg/cmake to find the SDK in the first place.)

I discovered when trying to run `vcvarsall.bat` that it was failing because it did not have permission to run `reg` (command-line registry lookup) to find the path to the SDK. By running `vcvarsall.bat` from a powershell that I run as Administrator, it works.

My current theory is that vcpkg/cmake is also failing to find the Windows SDK because it can't run `reg`, although in this case it's harder because, as far as I can see, the actual calls to `reg` must be baked into vcpkg.exe or cmake.exe. Running vcpkg from a powershell that I run as Administrator does *not* work in this case. (I don't know the rules for how admin permissions are or are not transferred to child processes, whether this differs between `.bat` and `.exe`, etc.)

I've tried playing with various cmake/environment variables. I've added these two lines to my triplet file (`x86-windows.cmake`):

```
set(VCPKG_ENV_PASSTHROUGH 1)
set(CMAKE_WINDOWS_KITS_10_DIR "C:\\Program Files (x86)\\Windows Kits\\10")
```

And I've also defined these environment (`env:`) variables:

```
CMAKE_WINDOWS_KITS_10_DIR      C:\Program Files (x86)\Windows Kits\10
VCPKG_ENV_PASSTHROUGH          1
```

None of that makes a difference.

I've got ahead and put below the output I get in `CMakeError.log`. As I mentioned above, since it can't find the SDK, I've tried to make the build work by putting everything it needs in `C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x86`. This is currently resulting in a linker error because of a x86/x64 mismatch. That's a secondary problem that I don't understand. I see that it has `/machine:x64` in both command lines below, and I don't know where that's coming from because I'm using the triplet file `x86-windows.cmake`. So potentially if I could fix that, I would have a workaround. But I'd prefer to have it use the SDK tool folder directly, if that's the way it should work.

Console
```
PS C:\Users\michael.katz\Documents> .\vcpkg\vcpkg.exe install shapelib
warning: Starting with the September 2023 release, the default triplet for vcpkg libraries will change from x86-windows to the detected host triplet (x64-windows). To resolve this message, add --triplet x86-windows to keep the same behavior.
Computing installation plan...
The following packages will be built and installed:
    shapelib[core]:x86-windows -> 1.5.0#4
Detecting compiler hash for triplet x86-windows...
Restored 0 package(s) from C:\Users\michael.katz\AppData\Local\vcpkg\archives in 339 us. Use --debug to see more details.
Installing 1/1 shapelib:x86-windows...
Building shapelib[core]:x86-windows...
-- Using cached shapelib-1.5.0.zip.
-- Cleaning sources at C:/Users/michael.katz/Documents/vcpkg/buildtrees/shapelib/src/shapelib-1-13b86997b8.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/Users/michael.katz/Documents/vcpkg/downloads/shapelib-1.5.0.zip
-- Applying patch option-build-test.patch
-- Using source at C:/Users/michael.katz/Documents/vcpkg/buildtrees/shapelib/src/shapelib-1-13b86997b8.clean
-- Found external ninja('1.11.0').
-- Configuring x86-windows
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: "C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -v
    Working Directory: C:/Users/michael.katz/Documents/vcpkg/buildtrees/shapelib/x86-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      C:\Users\michael.katz\Documents\vcpkg\buildtrees\shapelib\config-x86-windows-rel-CMakeCache.txt.log
      C:\Users\michael.katz\Documents\vcpkg\buildtrees\shapelib\config-x86-windows-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:248 (vcpkg_execute_required_process)
  ports/shapelib/portfile.cmake:17 (vcpkg_cmake_configure)
  scripts/ports.cmake:147 (include)


error: building shapelib:x86-windows failed with: BUILD_FAILED
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
    https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+shapelib
You can submit a new issue at:
    https://github.com/microsoft/vcpkg/issues/new?title=[shapelib]+Build+error&body=Copy+issue+body+from+C%3A%2FUsers%2Fmichael.katz%2FDocuments%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md
```

CMakeError.log
```
Detecting C compiler ABI info failed to compile with the following output:
Change Dir: C:/Users/michael.katz/Documents/vcpkg/buildtrees/shapelib/x86-windows-rel/CMakeFiles/CMakeScratch/TryCompile-nqlk2e

Run Build Command(s):C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe cmTC_571b9 && [1/2] Building C object CMakeFiles\cmTC_571b9.dir\CMakeCCompilerABI.c.obj

[2/2] Linking C executable cmTC_571b9.exe

FAILED: cmTC_571b9.exe 

cmd.exe /C "cd . && C:\Users\michael.katz\Documents\vcpkg\downloads\tools\cmake-3.25.1-windows\cmake-3.25.1-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=CMakeFiles\cmTC_571b9.dir --rc=C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x86\rc.exe --mt=C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x86\mt.exe --manifests  -- C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmTC_571b9.dir\CMakeCCompilerABI.c.obj  /out:cmTC_571b9.exe /implib:cmTC_571b9.lib /pdb:cmTC_571b9.pdb /version:0.0 /machine:x64  /debug /INCREMENTAL /subsystem:console  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK Pass 1: command "C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmTC_571b9.dir\CMakeCCompilerABI.c.obj /out:cmTC_571b9.exe /implib:cmTC_571b9.lib /pdb:cmTC_571b9.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\cmTC_571b9.dir/intermediate.manifest CMakeFiles\cmTC_571b9.dir/manifest.res" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'kernel32.lib'


ninja: build stopped: subcommand failed.





Determining if the C compiler works failed with the following output:
Change Dir: C:/Users/michael.katz/Documents/vcpkg/buildtrees/shapelib/x86-windows-rel/CMakeFiles/CMakeScratch/TryCompile-inujnz

Run Build Command(s):C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe cmTC_0f84e && [1/2] Building C object CMakeFiles\cmTC_0f84e.dir\testCCompiler.c.obj

[2/2] Linking C executable cmTC_0f84e.exe

FAILED: cmTC_0f84e.exe 

cmd.exe /C "cd . && C:\Users\michael.katz\Documents\vcpkg\downloads\tools\cmake-3.25.1-windows\cmake-3.25.1-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=CMakeFiles\cmTC_0f84e.dir --rc=C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x86\rc.exe --mt=C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x86\mt.exe --manifests  -- C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmTC_0f84e.dir\testCCompiler.c.obj  /out:cmTC_0f84e.exe /implib:cmTC_0f84e.lib /pdb:cmTC_0f84e.pdb /version:0.0 /machine:x64  /debug /INCREMENTAL /subsystem:console  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK Pass 1: command "C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmTC_0f84e.dir\testCCompiler.c.obj /out:cmTC_0f84e.exe /implib:cmTC_0f84e.lib /pdb:cmTC_0f84e.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\cmTC_0f84e.dir/intermediate.manifest CMakeFiles\cmTC_0f84e.dir/manifest.res" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'kernel32.lib'


ninja: build stopped: subcommand failed.
```

CMakeOutput.log
```
The target system is: Windows - 10.0.19044 - x86
The host system is: Windows - 10.0.19044 - AMD64
Compiling the C compiler identification source file "CMakeCCompilerId.c" succeeded.
Compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe 
Build flags: ;/nologo;/DWIN32;/D_WINDOWS;/W3;/utf-8;/MP;
Id flags: -c 

The output was:
0
CMakeCCompilerId.c


Compilation of the C compiler identification source "CMakeCCompilerId.c" produced "CMakeCCompilerId.obj"

The C compiler identification is MSVC, found in "C:/Users/michael.katz/Documents/vcpkg/buildtrees/shapelib/x86-windows-rel/CMakeFiles/3.25.1/CompilerIdC/CMakeCCompilerId.obj"
```

**Comment by autoantwort** [comment]
Maybe your installation is screwed. Maybe consider reinstalling the windows SDK and VS or even windows

**Reply by michaeladamkatz** [reply]
Well, I had a coworker follow my minimal steps (just install vcpkg and try to install a package) and she got the same error. We both use MSVC daily without problem.

**Reply by autoantwort** [reply]
So clean windows+ Vs + Windows SDK and then vcpkg does not work?

**Reply by michaeladamkatz** [reply]
Well, as I say, everything works on my personal computer (Windows 11 home). The work computers used by me and my coworker (Windows 10 Enterprise) have various permissions and configurations set by our employer and can't be changed. But otherwise yes everything is "clean" -- and we use Visual Studio and many other tools regularly without a problem. As I say, the one permissions difference that I think might be relevant is that running the reg command requires administrator privilege on the work computers but not on my personal computer.

**Reply by autoantwort** [reply]
https://gitlab.kitware.com/cmake/cmake/-/issues/17789
The registry is used to detect the windows SDK. So if You need admin rights to access the registry cmake maybe also needs them

**Reply by dg0yt** [reply]
But *reading* the relevant keys, or running `reg.exe` to query the keys, shouldn't require admin privilges.

Can you run `reg.exe` at all? Banning `reg.exe` wouldn't be a sound security measure on a machine which is to run MSVS, but I wouldn't be surprised to see such policies.

**Reply by michaeladamkatz** [reply]
Yes, that's my leading theory so far. I'm trying to figure out if there's a way to run vcpkg with admin privileges in a way that it passes those privileges down to other scripts and tools and it invokes, including cmake. Alternatively, I'm trying to figure out if there's a way to hack one of the cmake scripts in `C:\Users\michael.katz\Documents\vcpkg\scripts` to force the path to the SDK.

**Reply by michaeladamkatz** [reply]
@dg0yt It seems that reg.exe does indeed require admin privileges:

```
PS C:\Users\michael.katz\Documents> reg
Program 'reg.exe' failed to run: Access is deniedAt line:1 char:1
+ reg
+ ~~~.
At line:1 char:1
+ reg
+ ~~~
    + CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
    + FullyQualifiedErrorId : NativeCommandFailed
```

But when I run it from an Admin powershell, it works. BUT... my worry/thought is that when I run vcpkg from an admin powershell, because it invokes a chain of scripts and .exe files, the privilege doesn't get down to those other tools.

Yeah, I don't know enough to know what other tools would rely on running reg.exe to read registry values. I saw a comment in the link given by @autoantwort above that vcpkg is just following the pattern of Visual Studio in how it finds the SDK. But I wonder if that happens only once during install when the installer already has admin rights?

**Reply by dg0yt** [reply]
I don't think you showed that `reg.exe` *requires admin privilege*. You only showed that it *is blocked under your unprivileged account*, at least when running from PS. (Did you try to run it in `cmd`?)

I don't think you should run any build steps with admin privilege. (`vcpkg integrate install` is not a build step.)

**Reply by michaeladamkatz** [reply]
`reg` also says "Access is denied." under `cmd`. Well, regardless of whether it requires admin privilege or not, I am not able to run it from my account. And when I run `vcpkg` from an admin powershell, it gives the same behavior so my assumption is that it drops back to my privileges when running subprocesses (like ninja or cmake). By the way, I also tried logging into Windows with my local admin account, hoping that anything that I ran, and its subprocesses, would have admin privilege. But I don't think it works that way. So I'm still stuck.

Do you have any ideas of how I could hack some of the cmake scripts to force it to work? I've been trying to find the place where it looks in the registry for the SDK path, but I fear that logic is baked into vcpkg.exe or cmake.exe. I was also trying to hack the call to msbuild hoping to pass the forced SDK path as a parameter, but from what I can see you can specify the version number but not the path.

Oh, and I was also excited that `C:\Users\michael.katz\Documents\vcpkg\buildtrees\shapelib\x64-windows-static-release-rel\CMakeCache.txt` had all the paths listed, and said I could edit it. But it seems to get overwritten with each `vcpkg` run, so that's not an option. :(

**Reply by autoantwort** [reply]
I think it is backed into cmake. Vcpkg exe does iirc nothing with the windkwos SDK. 

You can set the triplet variable additional cmake config params or so described by the documentation

**Reply by michaeladamkatz** [reply]
@autoantwort I did try using the cmake parameters, as described in the OP above. It sounded promising but it didn't work.

**Reply by autoantwort** [reply]
https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_cmake_configure_options

**Reply by michaeladamkatz** [reply]
@autoantwort Yep.

**Reply by autoantwort** [reply]
Have you tried https://cmake.org/cmake/help/latest/variable/CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION.html ?

**Reply by autoantwort** [reply]
So setting the CMAKE_WINDOWS_KITS_10_DIR env var (you also must instruct vcpkg to pass the env var to cmake)

**Reply by michaeladamkatz** [reply]
@autoantwort Right, that's what I was trying in the OP above, right?

**Reply by autoantwort** [reply]
No you set a cmake variable and not an environment variable

**Reply by michaeladamkatz** [reply]
It shows above that I also set them as environment variables.

**Reply by autoantwort** [reply]
Sorry you are right, didn't read properly. But vcpkg_env_passthrough should be a list

https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_env_passthrough

**Reply by michaeladamkatz** [reply]
Okay, thanks, that sounds promising, but I didn't get it to work. I defined the below in x64-windows.cmake, and defined the same as environment variables. I also tried quoting "CMAKE_WINDOWS_KITS_10_DIR" and that didn't work either.

```
set(CMAKE_WINDOWS_KITS_10_DIR "C:/Program Files (x86)/Windows Kits/10")
set(CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION 10.0.22621.0)
set(VCPKG_ENV_PASSTHROUGH CMAKE_WINDOWS_KITS_10_DIR)
```

Also tried:

```
set(CMAKE_WINDOWS_KITS_10_DIR "C:/Program Files (x86)/Windows Kits/10")
set(CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION 10)
set(VCPKG_ENV_PASSTHROUGH "CMAKE_WINDOWS_KITS_10_DIR,CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION")
```

**Reply by autoantwort** [reply]
They are no cmake variables. It must be:
```cmake 


set(ENV{CMAKE_WINDOWS_KITS_10_DIR} "C:/Program Files (x86)/Windows Kits/10")
set(VCPKG_CMAKE_CONFIGURE_OPTIONS "-DCMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION=10.0.22621.0")
set(VCPKG_ENV_PASSTHROUGH CMAKE_WINDOWS_KITS_10_DIR)

```
I haven't tested it but it at least should have to chance to work

**Reply by michaeladamkatz** [reply]
Still doesn't work. I also tried it with backslashes as the directory separator. And I kind of didn't expect it to work if it's only setting an environment variable of that name, because I had already set an environment variable of that name explicitly.

**Reply by autoantwort** [reply]
Yeah maybe ask the admin to allow registry access from cmake

**Reply by michaeladamkatz** [reply]
Okay, thanks for all your time.
